lib/monitor.c

```diff
@@ -0,0 +1,173 @@
+  //     and subsequently the watch gets deleted by ivykis: ivykis/src/iv_inotify.c:88
+  //   2- handle case when file is edited and contains a parse error, syslog-ng changes
+  //     file attributes and file becomes unreadable
+
+  //    - note : for case when file was deleted, syslog-ng would still attempt to reload,
+  //           which gives a helpful "file not found" error in logs -> but auto-reload will then be disabled
+  //           since there's no file to watch anymore. ``syslog-ng-ctl reload`` should be issued after file is restored
+
+  else if (event->mask & IN_IGNORED || event->mask & IN_ATTRIB)
+    {
+      iv_inotify_unregister(&(fm->inotify));
+      iv_inotify_watch_unregister(fm->w);
+      file_monitor_start(fm);
+      fm->main_loop_reload_config(fm->self);
+    }
+}
+
+FileMonitor *
+file_monitor_create(const gchar *file_path)
+{
+  FileMonitor *fm;
+  struct iv_inotify_watch *w;
+
+  fm = g_malloc0(sizeof(*fm));
+  w = g_malloc0(sizeof(*w));
+
+  if ((fm == NULL) | (w == NULL))
+    {
+      perror("malloc");
+      exit(EXIT_FAILURE);
+    }
+
+  fm->w = w;
+  fm->w->cookie = (void *) fm;
+  fm->file_path = file_path;
+  return fm;
+}
+
+void
+file_monitor_config_reload_callback_function(FileMonitor *fm, void (*mainloop_callback)(void *), gpointer *self)
+{
+  fm->main_loop_reload_config = mainloop_callback;
+  fm->self = self;
+}
+
+void
+file_monitor_start(FileMonitor *fm)
+{
+  IV_INOTIFY_INIT(&(fm->inotify));
+  iv_inotify_register(&(fm->inotify));
+
+  IV_INOTIFY_WATCH_INIT(fm->w);
+  fm->w->inotify = &(fm->inotify);
+  fm->w->mask = IN_ALL_EVENTS;
+  fm->w->pathname = fm->file_path;
+  iv_inotify_watch_register(fm->w);
+  fm->w->handler = handler;
+}
+
+#else
+
+#define TIMEOUT (5)
+
+static void
+read_file(const gchar *file_path, gchar **file_contents)
+{
+  GError *err = NULL;
+
+  if (!g_file_get_contents(file_path, file_contents, NULL, &err))
+    {
+      printf("auto-reload: cannot read configuration file: %s", err->message);
+      g_error_free(err);
+      err = NULL;
+      return;
+    }
+}
+
+static void
+handler(void *_fm)
+{
+  FileMonitor *fm = (FileMonitor *) _fm;
+  gchar *file_contents = NULL;
+
+  read_file(fm->file_path, &file_contents);
+  if(file_contents)
+    {
+      if (strcmp(fm->current_config, file_contents) != 0)
+        {
+          fm->main_loop_reload_config(fm->self);
+          read_file(fm->file_path, &(fm->current_config));
+        }
+      g_free(file_contents);
+    }
+
+  file_monitor_start(fm);
+}
+
+FileMonitor *
+file_monitor_create(const gchar *file_path)
+{
+  FileMonitor *fm;
+  fm = g_malloc0(sizeof(*fm));
+
+  if (fm == NULL)
+    {
+      perror("malloc");
+      exit(EXIT_FAILURE);
+    }
+
+  fm->file_path = file_path;
+  read_file(fm->file_path, &(fm->current_config));
+
+  return fm;
+}
+
+void
+file_monitor_config_reload_callback_function(FileMonitor *fm, void (*mainloop_callback)(void *), gpointer *self)
```

Comment:
This function is repeated because of the `#if #else #endif`, but it can be avoided, as this function is the same in either case.